kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+local constants = require "kong.constants"
+local clustering_utils = require "kong.clustering.utils"
+-- currently they are the same. But it's possible for we to drop support for old version of DP but keep support of CP
+local supported_services = require "kong.clustering.services.supported"
+local asked_services = require "kong.clustering.services.supported"
+local table_clear = require "table.clear"
+
+local time = ngx.time
+local var = ngx.var
+local log = ngx.log
+local ERR = ngx.ERR
+local NOTICE = ngx.NOTICE
+local _log_prefix = "[wrpc-clustering] "
+local table_concat = table.concat
+local lower = string.lower
+local empty_table = {}
+
+local pairs = pairs
+local ipairs = ipairs
+local type = type
+local error = error
+
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+
+-- it's so annoying that protobuf does not support map to array
+local function wrap_services(service)
+  local wrapped = {}
+  for name, version in pairs(service) do
+    wrapped[name] = {
+      service = version
+    }
+  end
+
+  return wrapped
+end
+
+local function unwrap_services(service)
+  local unwrapped = {}
+  for name, version in pairs(service) do
+    unwrapped[name] = version.service
+  end
+
+  return unwrapped
+end
+
+local _M = {}
+
+local function field_validate(tbl, field, typ)
+  local v = tbl
+  for i, ind in ipairs(field) do
+    if type(v) ~= "table" then
+      error("field '" .. table_concat(field, ".", 1, i-1) .. "' cannot be indexed with " .. ind)
+    end
+    v = v[ind]
+  end
+
+  local compare_typ = typ
+  if typ == "array" or typ == "object" then
+    compare_typ = "table"
+  end
+
+  if type(v) ~= compare_typ then
+    local field_name = table_concat(field, '.')
+    error("field \"" .. field_name .. "\" must be of type " .. typ)
+  end
+end
+
+local function verify_request(body)
+  for field, typ in pairs{
+    [{
+      "node",
+    }] = "object",
```

Comment:
Could we change this to `"table", -- object`, then we can simplify the code above it .